Add llc integration for LLVM IR

rhysd opened the pull request:
<!--
READ THIS: Before creating a pull request, please consider the following first.

* The most important thing you can do is write tests. Code without tests
  probably doesn't work, and will almost certainly stop working later on. Pull
  requests without tests probably won't be accepted, although there are some
  exceptions.
* Read the Contributing guide linked above first.
* If you are adding a new linter, remember to update the README.md file and
  doc/ale.txt first.
* If you add or modify a function for converting error lines into loclist items
  that ALE can work with, please add Vader tests for them. Look at existing
  tests in the test/handler directory, etc.
* If you add or modify a function for computing a command line string for
  running a command, please add Vader tests for that.
* Generally try and cover anything with Vader tests, although some things just
  can't be tested with Vader, or at least they can be hard to test. Consider
  breaking up your code so that some parts can be tested, and generally open up
  a discussion about it.
* Have fun!
-->

I added [llc](https://llvm.org/docs/CommandGuide/llc.html) support for linting LLVM IR code. I added implementation, documents and tests. I confirmed all tests I added were passing.
llc is an official compiler provided by LLVM. So I added it as a default linter when opening LLVM IR source file.

w0rp commented on ale_linters/llvm/llc.vim:
```diff
@@ -0,0 +1,31 @@
+function! ale_linters#llvm#llc#GetCommand(buffer) abort
+    return ale#Var(a:buffer, 'llvm_llc_command')
+    \   . ' -filetype=null -o='
+    \   . ale#Escape(g:ale#util#nul_file)
```
I just merged a pull request which used `/dev/null` on Windows for elm. Do you know of `nul` will actually work for `-o` on Windows? I wouldn't be surprised if yet another tool has special code for exactly the string `/dev/null`, but doesn't know what to do with `nul`.

rhysd replied:
I'm usually not using Windows. But I dug LLVM source code a bit. Filesystem support is implemented here and llc seems to use this.

https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Support/FileSystem.h

As long as I read the code, LLVM does not handle `/dev/null` specially. So I think `nul` is available but `/dev/null` is unavailable on Windows.

w0rp commented on autoload/ale/linter.vim:
```diff
@@ -25,6 +25,7 @@ let s:default_ale_linters = {
 \   'csh': ['shell'],
 \   'go': ['gofmt', 'golint', 'go vet'],
 \   'help': [],
+\   'llvm': ['llc'],
```
You can remove this line. Without any entries in the Dictionary here, ALE will run all available tools for linting. The only reason to specify something here is to _disable_ some linter by default.

rhysd replied:
I'll remove the line, thanks.

w0rp commented on doc/ale-llvm.txt:
```diff
@@ -0,0 +1,19 @@
+===============================================================================
+ALE LLVM Integration                                           *ale-llvm-options*
```
I can't tell if the tags are aligned to the right margin properly or not on GitHub, but they might not be. Make sure they are.

rhysd replied:
I did not notice that `===` line is longer than title. I'll align titles.

w0rp commented on ale_linters/llvm/llc.vim:
```diff
@@ -0,0 +1,31 @@
+" Author: rhysd <https://rhysd.github.io>
+" Description: Support for checking LLVM IR with llc
+
+call ale#Set('llvm_llc_command', 'llc')
```
Name this variable `_executable` instead, so it matches the naming convention used for all other linters.

w0rp commented on ale_linters/llvm/llc.vim:
```diff
@@ -0,0 +1,31 @@
+
+call ale#linter#Define('llvm', {
+\   'name': 'llc',
+\   'executable': g:ale_llvm_llc_command,
```
Write a `GetExecutable` function for this, use `exectuable_callback` instead, and use `ale#Var` for the variable. Then users will be able to configure the executable per-buffer, and while Vim is running. Call that function from `GetCommand`, and apply `ale#Escape` to the executable path in `GetCommand`, which could contain spaces and other troublesome characters.

rhysd replied:
I'll rename the variable and add `GetCommand` callback.

rhysd commented:
Thank you for detailed reviews @w0rp! I'll fix the points.

rhysd commented:
@w0rp I believe I've fixed all points. Could you confirm?

w0rp commented on ale_linters/llvm/llc.vim:
```diff
@@ -0,0 +1,35 @@
+endfunction
+
+function! ale_linters#llvm#llc#GetCommand(buffer) abort
+    return ale_linters#llvm#llc#GetExecutable(a:buffer)
```
Use `ale#Escape` on the executable here. Otherwise, paths with spaces and other characters won't work.

w0rp commented on test/command_callback/test_llc_command_callback.vader:
```diff
@@ -0,0 +1,24 @@
+
+Execute(llc command is customizable):
+  let cmd = ale_linters#llvm#llc#GetCommand(bufnr(''))[:3]
+  AssertEqual 'llc ', cmd
```
When testing the executable names, use `ale#Escape` on them here. This will make the tests pass in both Unix and Windows environments. (Windows escaping is very different, and may not add quotes at all in some cases.)

rhysd commented:
@w0rp Thank you for catching the point. I fixed implementation and tests using `ale#Escape`. And I added another test to check the path is escaped correctly.

w0rp commented on test/command_callback/test_llc_command_callback.vader:
```diff
@@ -0,0 +1,29 @@
+Execute(space can be contained in llc executable path):
+  let b:ale_llvm_llc_executable = '/path/space contained/llc'
+  let cmd = ale_linters#llvm#llc#GetCommand(bufnr(''))
+  AssertEqual stridx(cmd, ale#Escape('/path/space contained/llc')), 0
```
I will probably end up changing these tests later, as the failure message will print `-1 != 0`, which doesn't tell you much of anything.

rhysd replied:
You're correct. But how can I write a test to check the string is prefixed? Is there any way to specify assertion message? I'm not familiar with vader.vim. So could you please help to improve these tests?

rhysd replied:
Ah, I found that message can be specified. I'll improve message. https://github.com/junegunn/vader.vim#execute

w0rp commented on ale_linters/llvm/llc.vim:
```diff
@@ -0,0 +1,35 @@
+call ale#Set('llvm_llc_executable', 'llc')
+
+function! ale_linters#llvm#llc#GetExecutable(buffer) abort
+    return ale#Escape(ale#Var(a:buffer, 'llvm_llc_executable'))
```
Nope, don't escape this here. The return value of the callbacks for getting the executable must not be a escaped, but the executable should be escaped in the command callback.

rhysd replied:
Ah, I got it.

w0rp commented on test/command_callback/test_llc_command_callback.vader:
```diff
@@ -0,0 +1,29 @@
+  Restore
+
+Execute(llc command is customizable):
+  let cmd = ale_linters#llvm#llc#GetCommand(bufnr(''))
```
Test the return value of `GetExecutable` too, and ensure it is not escaped.

rhysd replied:
I'll do.

rhysd commented:
@w0rp I fixed the points in c7c1c8e and e179661. Could you check them?

w0rp commented on test/command_callback/test_llc_command_callback.vader:
```diff
@@ -0,0 +1,38 @@
+
+  runtime ale_linters/llvm/llc.vim
+
+  function! AssertHasPrefix(str, prefix) abort
```
You'll need to delete this function in the `After` block, with `delfunction`. Otherwise it will be accessible from other tests.

rhysd replied:
Fixed in 1e72204 :+1:

w0rp commented on test/command_callback/test_llc_command_callback.vader:
```diff
@@ -14,6 +14,7 @@ Before:
 After:
   unlet! g:ale_llvm_llc_executable
   unlet! b:ale_llvm_llc_executable
+  delfunction! AssertHasPrefix
```
`delfunction!` doesn't work. The command doesn't accept a bang. (`!`)

rhysd replied:
ah, it was my mistake. And `:Vader` did not report the error. I modified commit at 1e72204 and `push -f`ed again.

rhysd commented:
@w0rp Thank you for reviewing many points! :smile:
